### ISSUE TITLE
Panic consuming trace data from gRPC - runtime error: index out of range [17] with length 17

### ISSUE THREAD

**Issue by LouisStAmour** [grafana/agent#648]
I'm running `grafana/agent:v0.15.0` receiving data from `otel-collector` over gRPC where there might be an incompatibility - I'm not certain - between the opentelemetry collector 0.21.0 sources mentioned in the backtrace and the protocol supported by the nearly 0.28.0 version I'm running. 

(I'm running a custom build of otel collector, ghcr.io/louisstamour/opentelemetry-collector-contrib/otelcontribcol:azure-events-v0.28.pre.2021-06-03, which adds extra support for Azure App Insights, but I'll update that brach shortly to the official release of 0.28.0 from 4 days ago.)

Given the error happens within opentelemetry code, it's possible it's an upstream error that should be handled without a panic crashing everything. I'm not sure if it's possible to catch these kinds of panics and recover from them, but it does seem like a desirable behaviour! :D

For now, I've just told my agent container to restart locally, as this isn't a production configuration yet for me, just something I'm testing in docker-compose.

```
agent                     | ts=2021-06-13T07:27:23Z level=info msg="Consuming trace data" component=tempo tempo_config=default component_kind=processor component_type=spanmetrics component_name=spanmetrics
agent                     | panic: runtime error: index out of range [17] with length 17
agent                     | 
agent                     | goroutine 19999 [running]:
agent                     | github.com/open-telemetry/opentelemetry-collector-contrib/processor/spanmetricsprocessor.(*processorImp).updateLatencyMetrics(...)
agent                     | 	/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/processor/spanmetricsprocessor@v0.21.0/processor.go:313
agent                     | github.com/open-telemetry/opentelemetry-collector-contrib/processor/spanmetricsprocessor.(*processorImp).aggregateMetricsForSpan(0xc000c4c900, 0xc003c2a7e6, 0xa, 0xc001bfc2a0)
agent                     | 	/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/processor/spanmetricsprocessor@v0.21.0/processor.go:297 +0x47a
agent                     | github.com/open-telemetry/opentelemetry-collector-contrib/processor/spanmetricsprocessor.(*processorImp).aggregateMetricsForServiceSpans(0xc000c4c900, 0xc001495200, 0xc003c2a7e6, 0xa)
agent                     | 	/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/processor/spanmetricsprocessor@v0.21.0/processor.go:281 +0x5e
agent                     | github.com/open-telemetry/opentelemetry-collector-contrib/processor/spanmetricsprocessor.(*processorImp).aggregateMetrics(0xc000c4c900, 0xc00304bbc0)
agent                     | 	/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/processor/spanmetricsprocessor@v0.21.0/processor.go:270 +0xc8
agent                     | github.com/open-telemetry/opentelemetry-collector-contrib/processor/spanmetricsprocessor.(*processorImp).ConsumeTraces(0xc000c4c900, 0x42645b0, 0xc002833290, 0xc00304bbc0, 0x3a973a5, 0x4)
agent                     | 	/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/processor/spanmetricsprocessor@v0.21.0/processor.go:176 +0x71
agent                     | go.opentelemetry.io/collector/receiver/otlpreceiver/trace.(*Receiver).sendToNextConsumer(0xc00027f9a0, 0x42645b0, 0xc002833290, 0xc00304bbc0, 0x3a973a5, 0x4)
agent                     | 	/go/pkg/mod/go.opentelemetry.io/collector@v0.21.0/receiver/otlpreceiver/trace/otlp.go:99 +0x146
agent                     | go.opentelemetry.io/collector/receiver/otlpreceiver/trace.(*Receiver).Export(0xc00027f9a0, 0x42645b0, 0xc0028331d0, 0xc00304bb60, 0xc00027f9a0, 0xc0028331d0, 0xc002593ba0)
agent                     | 	/go/pkg/mod/go.opentelemetry.io/collector@v0.21.0/receiver/otlpreceiver/trace/otlp.go:80 +0x179
agent                     | go.opentelemetry.io/collector/internal/data/protogen/collector/trace/v1._TraceService_Export_Handler(0x346b180, 0xc00027f9a0, 0x42645b0, 0xc0028331d0, 0xc005488540, 0x0, 0x42645b0, 0xc0028331d0, 0xc003186000, 0x5fe3)
agent                     | 	/go/pkg/mod/go.opentelemetry.io/collector@v0.21.0/internal/data/protogen/collector/trace/v1/trace_service.pb.go:209 +0x214
agent                     | google.golang.org/grpc.(*Server).processUnaryRPC(0xc000794e00, 0x42952b8, 0xc0031ef680, 0xc004c579e0, 0xc000b3a840, 0x5a76020, 0x0, 0x0, 0x0)
agent                     | 	/go/pkg/mod/google.golang.org/grpc@v1.37.0/server.go:1217 +0x52b
agent                     | google.golang.org/grpc.(*Server).handleStream(0xc000794e00, 0x42952b8, 0xc0031ef680, 0xc004c579e0, 0x0)
agent                     | 	/go/pkg/mod/google.golang.org/grpc@v1.37.0/server.go:1540 +0xd0c
agent                     | google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc002785620, 0xc000794e00, 0x42952b8, 0xc0031ef680, 0xc004c579e0)
agent                     | 	/go/pkg/mod/google.golang.org/grpc@v1.37.0/server.go:878 +0xab
agent                     | created by google.golang.org/grpc.(*Server).serveStreams.func1
agent                     | 	/go/pkg/mod/google.golang.org/grpc@v1.37.0/server.go:876 +0x1fd

```

**Comment by mapno** [comment]
Hi @LouisStAmour, thanks for opening an issue!

Seeing the stack trace of the panic, it all points to being an issue in the [processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/v0.21.0/processor/spanmetricsprocessor). It seems to be an issue with the histogram buckets, as if there was no _catch all_ bucket, but the processor already considers this [case](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.21.0/processor/spanmetricsprocessor/processor.go#L95-L98).

Can you share your configuration for the Grafana Agent? If we find the panic to come from the processor, we can move the issue to [otel's repo](https://github.com/open-telemetry/opentelemetry-collector-contrib).

**Comment by LouisStAmour** [comment]
Here's what my agent.yaml looks like - it's based on the stock Docker-Compose but with a few modifications:

```
server:
  log_level: info
  http_listen_port: 12345
prometheus:
  wal_directory: /tmp/agent
  global:
    scrape_interval: 5s
    remote_write:
      - url: http://cortex:9009/api/prom/push
  configs:
    - name: default
      scrape_configs:
      - job_name: cortex
        static_configs:
          - targets: ['localhost:9009']
      - job_name: loki
        static_configs:
          - targets: ['localhost:3100']
      - job_name: otel-collector
        static_configs:
          - targets: ['otel-collector:8888', 'otel-collector:8889']
      - job_name: celery-exporter
        static_configs:
          - targets: ['celery-exporter:6060']
    - name: tempo
      host_filter: false
      remote_write:
        - url: http://cortex:9009/api/prom/push
loki:
  positions_directory: /tmp/loki-positions
  configs:
  - name: default
    clients:
      - url: http://loki:3100/loki/api/v1/push
    scrape_configs:
    - job_name: system
      static_configs:
      - targets: ['localhost']
        labels:
          job: varlogs
          __path__: /var/log/*log
tempo:
  configs:
  - name: default
    receivers:
      otlp:
        protocols:
          grpc:
    remote_write:
      - endpoint: tempo:55680
        insecure: true
    batch:
      timeout: 5s
      send_batch_size: 100
    automatic_logging:
      backend: loki
      loki_name: default
      spans: true
      processes: true
      roots: true
    spanmetrics:
      prom_instance: tempo
integrations:
  agent:
    enabled: true
  node_exporter:
    enabled: true
    rootfs_path: /host/root
    sysfs_path: /host/sys
    procfs_path: /host/proc
  process_exporter:
    enabled: true
    process_names:
      - name: "{{.Comm}}"
        cmdline:
        - '.+'
  postgres_exporter:
    enabled: true
    data_source_names:
      - "${POSTGRES_DATABASE_URI}"
  redis_exporter:
    enabled: true
    redis_addr: "${REDIS_ADDR}"
    redis_password: "${REDIS_PASSWORD}"
  statsd_exporter:
    enabled: true
```

**Comment by mapno** [comment]
Everything looks good. I can't figure out what might be the issue. I wasn't able to reproduce the problem in the processor. Is this panic frequently happening?

**Comment by LouisStAmour** [comment]
I had a feeling it would be hard to reproduce. It’s a local development environment for now, we’re not yet deploying Agent to Kubernetes. I’ll see if I can load the agent’s crash logs into Loki somehow so I can better trace any crashes as log lines fly fast and furious. If I could, I would maybe have the server output its own OTLP on panic to avoid having to try and capture logs. Extra points if it can figure out how to mark the data as dropped and recover, moving on to the next datapoint and recording the panic. For now the logs are hard to follow, I only noticed because when the Agent Docker Container wasn’t set to restart automatically, it died without a trace (literally, no traces, lol), which is inconsistent with how Otel Collector is supposed to handle errors (silently recover and count them, etc.) but is to be expected given this is Go, after all, with intentionally minimal error handling.
